sklearn/decomposition/dict_learning.py

```diff
@@ -373,11 +373,12 @@ def _update_dict(dictionary, Y, code, verbose=False, return_r2=False,
     n_components = len(code)
     n_features = Y.shape[0]
     random_state = check_random_state(random_state)
-    # Residuals, computed 'in-place' for efficiency
-    R = -np.dot(dictionary, code)
-    R += Y
-    R = np.asfortranarray(R)
+    # Get BLAS functions
+    gemm, = linalg.get_blas_funcs(('gemm',), (dictionary, code, Y))
     ger, = linalg.get_blas_funcs(('ger',), (dictionary, code))
+    # Residuals, computed with BLAS for speed and efficiency
+    # R <- -1.0 * U * V^T + 1.0 * Y
+    R = gemm(-1.0, dictionary, code, 1.0, Y)
```

Comment:
This produces a Fortran array I presume?

Reply:
Yep.

Reply:
Maybe it would be worth making it explicit that we expect a fortran array in the comment.

Reply:
Interestingly, the SciPy linalg functions correctly handle C ordered arrays as input.

```python
In [1]: import numpy as np

In [2]: import scipy.linalg as linalg

In [3]: np.random.seed(0)

In [4]: a = np.random.random((2, 3))

In [5]: b = np.random.random((3, 4))

In [6]: c = np.random.random((2, 4))

In [7]: gemm, = linalg.get_blas_funcs(('gemm',), (a, b, c))

In [8]: gemm(1.0, a, b, 1.0, c)
Out[8]: 
array([[1.62736178, 1.79020759, 1.92593582, 2.17344607],
       [1.08121316, 1.54678646, 0.89707181, 1.77876957]])

In [9]: gemm(1.0, a, b, 1.0, c).flags
Out[9]: 
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
  UPDATEIFCOPY : False
```

That said, we appear to already be forcing `dictionary` and `code` to Fortran ordered arrays anywhere `_update_dict` is called.